linked_list/linked_list.py

```diff
@@ -9,109 +9,205 @@ def __init__(self, value, next_node = None):
     def get_at_index(self, index):
-        pass
+        if (self.size + 1) < index:
+            return None
+
+        counter = 0
+        current = self.head
+        while current:
+            if counter == index:
+                return current.value
+            current = current.next
+            counter += 1
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+
+        current = self.head
+
+        while current.next:
+            current = current.next
+        return current.value
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        if self.size == 0:
+            self.add_first(value)
+        else:
+            new_node = Node(value)
+            current = self.head
+
+            while current.next:
+                current = current.next
+            current.next = new_node
+            self.size += 1
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        if self.head == None:
+            return None
+
+        max = self.head.value
+        current = self.head
+
+        while current:
+            if current.value > max:
+                max = current.value
+            current = current.next
+
+        return max
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(self, value):
-        pass
+        if self.size == 0:
+            return None
+
+        current = self.head
+        if current.value == value:
+            self.head = current.next
+
+        while current.next:
+            if current.next.value == value:
+                current.next = current.next.next
+            else:
+                current = current.next
+        self.size -= 1
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit(self):
         helper_list = []
         current = self.head
 
         while current:
             helper_list.append(str(current.value))
             current = current.next
-        
+
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def reverse(self):

```

Comment:
👍 However the space complexity here is O(1), as you're not creating a new list, just reversing in place.